src/common/intl.cpp

```diff
@@ -270,26 +273,35 @@ bool wxLocale::Init(const wxString& name,
                     wxS("no locale to set in wxLocale::Init()") );
     }
 
-    if ( const wxLanguageInfo* langInfo = FindLanguageInfo(szLocale) )
+    int languageId = wxLANGUAGE_UNKNOWN;
+    wxLocaleIdent localeId = wxLocaleIdent::FromTag(szLocale);
+    if ( const wxLanguageInfo* langInfo = wxUILocale::FindLanguageInfo(localeId) )
     {
         // Prefer to use Init(wxLanguage) overload if possible as it will
         // correctly set our m_language and also set the locale correctly under
         // MSW, where just calling wxSetLocale() as we do below is not enough.
         //
-        // However don't do it if the parameters are incompatible with this
-        // language, e.g. if we are called with something like ("French", "de")
-        // to use French locale but German translations: this seems unlikely to
-        // happen but, in principle, it could.
-        if ( langInfo->CanonicalName.StartsWith(shortName) )
+        // However don't do it if
+        //   1. the locale is not empty, or
+        //   2. the parameters are incompatible with this language,
+        //      e.g. if we are called with something like ("French", "de")
+        //      to use French locale but German translations:
+        //      this seems unlikely to happen but, in principle, it could.
+        if ( locale.empty() && langInfo->CanonicalName.StartsWith(shortName) )
         {
             return Init(langInfo->Language,
                         bLoadDefault ? wxLOCALE_LOAD_DEFAULT : 0);
         }
+        else
+        {
+            strName = langInfo->Description;
+            strShort = langInfo->CanonicalRef.empty() ? langInfo->CanonicalName : langInfo->CanonicalRef;
```

Comment:
We probably should add `wxLanguageInfo::GetCanonicalRefOrName()` (except with a better name if you can come up with one) instead of writing the same not particularly clear `CanonicalRef.empty() ? CanonicalName : CanonicalRef` expression in several places.

Reply:
> We probably should add `wxLanguageInfo::GetCanonicalRefOrName()` (except with a better name if you can come up with one) instead of writing the same not particularly clear `CanonicalRef.empty() ? CanonicalName : CanonicalRef` expression in several places.

Yes, makes sense. At first glance, I thought `wxLanguageInfo::GetCanonicalWithRegion()` might be a good name, because usually that's what you get. However, there are several exceptions (mostly for exotic languages). Nevertheless, this method name seems to be more suitable.